clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordBatch.java

```diff
@@ -133,7 +133,7 @@ public boolean maybeExpire(int requestTimeout, long now, long lingerMs) {
         if ((this.records.isFull() && requestTimeout < (now - this.lastAppendTime)) || requestTimeout < (now - (this.lastAttemptMs + lingerMs))) {
             expire = true;
             this.records.close();
-            this.done(-1L, new TimeoutException("Batch Expired"));
+            this.done(-1L, new TimeoutException("Batch containing " + recordCount + " record(s) expired due to timeout while requesting metadata from brokers for " + topicPartition.topic() + "-" + topicPartition.partition()));
```

Comment:
Tiny nitpick: `TopicPartition.toString` does what you are doing manually here, so you could simplify it by just saying:

```java
"Batch containing " + recordCount + " record(s) expired due to timeout while requesting metadata from brokers for " + topicPartition
```